Add support for automatically retrying failed requests

nbrooke opened the pull request:
Also includes some associated cleanup, including:
- Removing the NotificationCenter events for request completion.
- The names and payload of some of the request related LogEvents have changed, generally to include more data on requests (so that can be used instead of the now-removed NotificationCenter events in almost all cases)
- LogEvent.requestCompleted is no longer sent for requests that fail based on non 2xx HTTP codes, just LogEvent.requestFailure.
- LogEvent.requestCompleted renamed to LogEvent.requestSuccess
- Some LogDestination calls would happen on the networking thread, those are now bridged back to the main thread.
- Default logging now logs URLs on request completion, and logs HTTP status codes for HTTP failures
- Change RequestIdentifier to be opaque (it now contains a different unique identifier, since the NSURLSessionTask identifier no longer uniquely identifies a request when retries can happen).
- Centralizing error handling a lot more in the actual network request (rather than logging and calling the completion callback in multiple places, all error handling now funnels through a single catch handler)

brendanlensink commented on Netable/Netable/DelayedOperations.swift:
```diff
@@ -0,0 +1,40 @@
+// Utility class for scheduling future operations that can easily be identified and cancelled later
+// Used to hold in flight requests that are waiting for a retry.
+internal class DelayedOperations {
+    private var operations: [(timer: Timer, id: String)] = []
```
Is there any difference here between using a tuple or struct here?
I'm mostly just curious, I don't think it matters either way since this is an internal class.

nbrooke replied:
Less typing, since it's only used in one place, but not much difference other than that.

brendanlensink commented on Netable/Netable/LogDestination.swift:
```diff
@@ -10,21 +10,33 @@ import Foundation
-    case requestStarted(urlString: String, method: HTTPMethod, headers: [String: Any], params: [String: Any]?)
+    case requestStarted(request: RequestInfo)
+
+    /// Request has sucessfully completed. Note: taskTime only cover the time it took the current network request to run, in retry scenarios the time for the whole request may be longer.
```
```suggestion
    /// Request has sucessfully completed. 
    /// Note: taskTime only covers the time it took the current network request to run, in retry scenarios the time for the whole request may be longer.
```

brendanlensink commented on Netable/Netable/Netable.swift:
```diff
@@ -73,82 +80,93 @@ open class Netable {
-                    object: self,
-                    userInfo: userInfo
-                )
+        logDestination.log(event: .requestStarted(request: requestInfo))
```
The changelog makes it seem like _all_ log events are now run on the main thread, but isn't it possible log events like this aren't?

Why is it important log events are run on the main thread? If it is important, is it maybe worth baking into `logDestination.log`?

nbrooke replied:
Yeah, I guess it's only running everything on the main thread IF all of the incoming calls are also happening on the main thread, which in practice they probably always are, but we shouldn't be depending on.

Main reason that I think they should all be on the main thread is because of the idea that stuff that was done with notifications could / should be done with logs instead. Once the logging is potentially doing something a little more complicated than just formatting the text and passing on to another (presumably) thread safe print function, it's a little risky to just have those coming back on whatever random thread.

The notification already had this problem (any code that used it needed to dispatch to the main thread), it seems like a good idea to just do that by default. Though maybe we need a little more sophisticated approach (always dispatch to main, even for calls we think are probably already on main, specify a user defined queue, etc.).

I don't know that it needs to be fixed to merge this, since I think the current behaviour is an improvement in practice, but we should probably file a bug to clean this up further.

brendanlensink commented on Netable/Netable/Netable.swift:
```diff
@@ -73,82 +80,93 @@ open class Netable {
+
+                if !isCancellation && retriesLeft > 0 && retryConfiguration.enabled && retryConfiguration.errors.shouldRetry(netableError) {
+                    mainThreadLog(.requestRetrying(request: requestInfo, taskTime: time, error: netableError))
+                    DispatchQueue.main.async {
```
Does it make more sense to move this main thread check into `delay` so it doesn't need to be copied over if we called it somewhere else?

nbrooke replied:
I went back and forth on this, the main reason I left it here is that the limitation is NOT that the DelayedOperations class REQUIRES it to be on the main thread, it just requires it to be on a thread with an active RunLoop (otherwise you can't schedule the timer) which the network thread(s) seem not to have. So although I don't think it matters in practice, it seemed more conceptually right to leave DelayedOperations able to work on any thread (with a RunLoop) and have the networking code force it onto the main thread because it doesn't have a RunLoop.

The fact that it confused you does suggest that something needs to be clearer, I think what I'll probably do it just switch it to use a DispatchSourceTimer instead, that should keep it as general purpose, without requiring any calls to be bridged to the main thread.

nbrooke commented:
Made the following changes since the last review:
  - Changed the defaults to be more in line with the Slack discussion
  - Added Brendan's comment typo fixes
  - Made DelayedOperations (I think) thread safe and removed the implicit dependancy on being called on a thread with a valid RunLoop. 
  - Went ahead and just made it so that we do implicitly bridge all log messages to the main thread. We should probably at some point have a way to turn that off. Even though it's a safer default, for plain old logging (which should itself be thread safe), it is an efficiency loss, and might in a few rare cases cause some logs to appear in the wrong order (i.e. Netable messages out of order relative to system logs also originating on the network thread) or be missing (if, for example, a crash happens between the event happening and the next time the main thread runs).

nbrooke commented:
I haven't bumped the version / triggered a release yet, since I'm not totally convinced it's not still going to need some other changes (either to support the stuff I'm doing in Spies, or to maybe address the affect on the project that is using the notification support). I'm just integrating directly from master on Spies, @JeremyChiang you should maybe try pulling the latest in at some point and see if you can get it working reasonably using the LogDestination instead of notification, if there are any changes that need to be made there.
